File: p089m/gray_code.js
/**
 * @param {number} n
 * @return {number[]}
 */
var grayCode = function (n) {
  let ans = [0];
  for (let i = 0; i < n; ++i) {
    for (let j = ans.length - 1; j >= 0; --j) {
      // '1' + reversed list of (n-1)
      ans.push(ans[j] | (1 << i));
    }
  }
  return ans;
};
// TEST
[0, 1, 2, 3, 4].forEach((n) => {
  console.log('Gray code of', n, '->');
  console.log('  ', grayCode(n));
});

File: p089m/gray_code.py
from typing import List


class Solution:
    def grayCode(self, n: int) -> List[int]:
        ans = [0]
        for i in range(n):
            ans += [n | (1 << i) for n in ans[::-1]]
        return ans


for n in range(5):
    sol = Solution()
    print("Gray code of", n, "->", sol.grayCode(n))
